### DISCUSSION TITLE
[Enhance-MetricGAN] Where does the checkpoint in the last step come from?

### DISCUSSION THREAD

**Discussion by a897456** [speechbrain/speechbrain#2611] [General]
speechbrain.utils.train_logger - Epoch: 750 - train loss: 1.12 - valid SI-SNR: 2.80, valid pesq: 2.81, valid stoi: 8.40e-01
speechbrain.utils.checkpoints - Saved an end-of-epoch checkpoint in results\MetricGAN\4236\save\CKPT+2024-07-14+22-47-58+00
speechbrain.utils.checkpoints - Deleted checkpoint in results\MetricGAN\4236\save\CKPT+2024-07-14+22-41-47+00
speechbrain.utils.epoch_loop - Going into epoch 751
Discriminator training by current data...
100%|██████████| 100/100 [00:22<00:00,  4.54it/s, train_loss=0.00111]
Discriminator training by historical data...
100%|██████████| 15020/15020 [04:24<00:00, 56.80it/s, train_loss=0.00114]
Discriminator training by current data again...
100%|██████████| 100/100 [00:14<00:00,  6.69it/s, train_loss=0.00122]
Generator training by current data...
100%|██████████| 100/100 [00:17<00:00,  5.70it/s, train_loss=1.11]
  0%|          | 0/39 [00:00<?, ?it/s]Avg G loss: 1.107
Avg D loss: 0.001
100%|██████████| 39/39 [00:52<00:00,  1.35s/it]
speechbrain.utils.train_logger - Epoch: 751 - train loss: 1.11 - valid SI-SNR: 2.63, valid pesq: 2.81, valid stoi: 8.40e-01
speechbrain.utils.checkpoints - Saved an end-of-epoch checkpoint in results\MetricGAN\4236\save\CKPT+2024-07-14+22-54-13+00
speechbrain.utils.checkpoints - Deleted checkpoint in results\MetricGAN\4236\save\CKPT+2024-07-14+22-47-58+00
speechbrain.utils.checkpoints - Loading a checkpoint from results\MetricGAN\4236\save\CKPT+2024-07-14+17-26-33+00
100%|██████████| 824/824 [01:13<00:00, 11.22it/s]
speechbrain.utils.train_logger - Epoch loaded: 697 - test SI-SNR: 8.17, test pesq: 3.14, test stoi: 9.26e-01

###  Epoch loaded: 697？ why? where?

**Comment by asumagic** [comment]
It looks like it's failing to delete an older checkpoint, `CKPT+2024-07-14+17-26-33+00`.  
Seems like a bug (maybe a `/` vs `\` splitting bug due to windows..?).  
From what I'm seeing of the logs, `CKPT+2024-07-14+22-54-13+00` is the only checkpoint that should still exist.

Can you try moving `CKPT+2024-07-14+17-26-33+00` away then rerunning the training script to see if it now evaluates on the correct one?

**Reply by a897456** [reply]
> Can you try moving `CKPT+2024-07-14+17-26-33+00` away then rerunning the training script to see if it now evaluates on the correct one?
speechbrain.core - Beginning experiment!
speechbrain.core - Experiment folder: ./results/MetricGAN/4236
speechbrain.core - Info: max_grad_norm arg from hparam file is used
speechbrain.core - Gradscaler enabled: False. Using precision: fp32.
speechbrain.core - MetricGanBrain Model Statistics:
* Total Number of Trainable Parameters: 400.8M
* Total Number of Parameters: 400.8M
* Trainable Parameters represent 100.0000% of the total size.
speechbrain.utils.checkpoints - Loading a checkpoint from results\MetricGAN\4236\save\CKPT+2024-07-14+22-54-13+00
speechbrain.utils.checkpoints - Loading a checkpoint from results\MetricGAN\4236\save\CKPT+2024-07-14+22-54-13+00
100%|██████████| 824/824 [01:10<00:00, 11.72it/s]
speechbrain.utils.train_logger - Epoch loaded: 751 - test SI-SNR: 8.54, test pesq: 3.12, test stoi: 9.28e-01

**Reply by a897456** [reply]
I would like to know how the best epoch is chosen? Is epoch 751 the one with the highest PESQ score at present? There is a PESQ score of 3.14 for epoch 697, how can it be replicated?

**Reply by asumagic** [reply]
Ah, ok, after reading the recipe code it turns out I was wrong. The code **does** select a model, and it selects the highest PESQ _on the validation dataset_ (here, at epoch 697), which is why the test PESQ is slightly lower (probably within margin of error):

```yaml
target_metric: pesq
```

```python
self.checkpointer.save_and_keep_only(
    meta=stats, max_keys=[self.hparams.target_metric]
)
```

So the checkpoint file that I asked to move away is the model you should select as the best one.

But, realistically, the results are so close you're probably fine taking the checkpoint from epoch 751.

**Reply by a897456** [reply]
> Ah, ok, after reading the recipe code it turns out I was wrong. The code **does** select a model, and it selects the highest PESQ _on the validation dataset_ (here, at epoch 697), which is why the test PESQ is slightly lower (probably within margin of error):
![image](https://github.com/user-attachments/assets/0b86c038-750c-4af0-8ac8-bc265eb16775)

**Reply by asumagic** [reply]
The checkpointing for this recipe keeps the checkpoint with the highest _validation_ PESQ.

- At epoch **694**, it is `2.76`.
- At epoch **695**, it is `2.80`. Thus, it's normal that the checkpoint from **694** was removed since it's better than the PESQ of **694**.
- At epoch **696**, it is `2.81`. Here, it deleted the checkpoint from **695**, for the same reason.
- At epoch **697**, it is `2.83`. Here, it deleted the checkpoint from **696**, for the same reason.

The PESQ of `3.14` is achieved on the _test_ dataset. It could be entirely normal for this value to be significantly different from validation, but I am not familiar with this dataset.

SpeechBrain only computes the **test** PESQ at the _end_ of training, using the model that had the best **valid** PESQ.  
This is the correct thing to do when evaluating models.

**Reply by a897456** [reply]
> 
> The PESQ of `3.14` is achieved on the _test_ dataset. It could be entirely normal for this value to be significantly different from validation, but I am not familiar with this dataset.
> 
The PESQ score of 3.14 is the result I modified (by adding a set of loss). I did this training some time ago: the code remained unchanged and was directly trained. In the end, the PESQ score was only 3.07, but the author obtained a score of 3.15 in the paper and the log. So I don't know what went wrong.

> 
> SpeechBrain` only computes the test PESQ at the end of training, using the model that had the best valid PESQ. This is the correct thing to do when evaluating models.
> 
If I want to use this checkpoint for individual inference, is there any inference code? (For example, using the checkpoint provided by the author for inference to view the PESQ score.)

**Reply by a897456** [reply]
> > The PESQ of `3.14` is achieved on the _test_ dataset. It could be entirely normal for this value to be significantly different from validation, but I am not familiar with this dataset.
> 
> The PESQ score of 3.14 is the result I modified (by adding a set of loss). I did this training some time ago: the code remained unchanged and was directly trained. In the end, the PESQ score was only 3.07, but the author obtained a score of 3.15 in the paper and the log. So I don't know what went wrong.
> 
> > SpeechBrain` only computes the test PESQ at the end of training, using the model that had the best valid PESQ. This is the correct thing to do when evaluating models.
> 
> If I want to use this checkpoint for individual inference, is there any inference code? (For example, using the checkpoint provided by the author for inference to view the PESQ score.)

Hi@asumagic 
Can you help me identify where the problem lies? Thank you

**Reply by asumagic** [reply]
> The PESQ score of 3.14 is the result I modified (by adding a set of loss). I did this training some time ago: the code remained unchanged and was directly trained. In the end, the PESQ score was only 3.07, but the author obtained a score of 3.15 in the paper and the log. So I don't know what went wrong.

I'm not sure I fully understand what was attempted here, but I am not very familiar with this task, so I don't think I can help much. Hyperparameters like batch size can affect it if you've changed it, though.

> If I want to use this checkpoint for individual inference, is there any inference code? (For example, using the checkpoint provided by the author for inference to view the PESQ score.)

Usually you need to make a custom hyperparameters YAML for inference. This is a manual process. It is helpful to check how existing models on HF do it.

In this case, this uploaded model seems to give the same PESQ figure, which might be what you want: https://huggingface.co/speechbrain/metricgan-plus-voicebank

**Comment by a897456** [comment]
[log.txt](https://github.com/user-attachments/files/16233908/log.txt)
This is the log which may be able to help you analyze the problem for me. Thank you.
